### PR TITLE
proxy api

### PR DESCRIPTION
**Pull request by sagold**

### CODE REVIEW

File: src/Foxy.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            this.handlers.push(handler);
+        }
+        this.handlers.splice(index, 0, handler);
+    }
+
+    removeHandler(handler: Handler): void {
+        this.handlers = this.handlers.filter(h => h !== handler);
+    }
+
+    getImageURL(request: Request): Promise<string> {
```
Comment (benurb) [inline src/Foxy.ts]: You might want to take a look at ES2015 proxies https://stackoverflow.com/a/42244695

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+function loadImageInfo(url: string): Promise<ImageInfo> {
+    const result = {
+        image: new Image(),
+        type: "",
+        width: 0,
+        height: 0
+    };
+
+    return fetch(url)
```
Comment (benurb) [inline src/utils.ts]: It would be cool to be able to pass in fetch options e.g. for custom headers.

File: src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    const result = {
+        type: "",
+        width: 0,
+        height: 0,
+        video: document.createElement("video"),
+        duration: 0,
+        poster: ""
+    };
+
+    return fetch(url)
```
Comment (benurb) [inline src/utils.ts]: Same as above

File: README.md [hunk 4]
```diff
@@ -1,2 +1,84 @@
...
+
+## Usage
+
+```js
+import { Foxy, utils } from "foxy";
+
+const proxy = new Foxy({
+  handlers: [
+    {
+      use({ url }) {
```
[the remark this answers is not in the document]
Comment (sagold) [reply README.md]: Its the whole request-object, passed to all handler-methods the same way. This guarantees flexibility to register really weird handlers, that may check any additional properties on the input-request (or event use different properties as url)

File: package.json [hunk 5]
```diff
@@ -1,10 +1,15 @@
...
-  "main": "dist/index.js",
+  "main": "dist/foxy.js",
+  "types": "ts/src/index.d.ts",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "build": "rm -rf dev; webpack",
+    "dist": "rm -rf dist; NODE_ENV=production webpack -p",
+    "lint": "eslint test/**/* src/**/*",
+    "test": "TS_NODE_PROJECT=./test/tsconfig.json mocha -r ts-node/register --require ./test/setup.ts 'test/**/*.test.ts'",
+    "prepublish": "npm run dist"
```
[the remark this answers is not in the document]
Comment (sagold) [reply package.json]: yes, i do!

File: src/utils.ts (hunk 2 above)
Comment (sagold) [reply src/utils.ts]: Yes

File: src/utils.ts (hunk 3 above)
Comment (sagold) [reply src/utils.ts]: Done

**Pull request merged**